### PR TITLE
feat(download): choose how missing measurements are exported

### LINKED ISSUE #102: Download: Allow choosing how missing  phenotypes are represented in a dropdown
When downloading phenotypes, give users the option on how they want missing values to be encoded. Possible values include:

- NA
- Empty String
- Period

### PR DESCRIPTION
**Pull request by ktmeaton**
Description 
-----------------------------------------------------

This PR allows the user to customize how they would like missing phenotype measurements to be displayed when exporting. Since I was changing forms, I added some small style changes to accomodate the new UI dimensions.

- Resolves #102
- Resolves #178 

## Tests To Be Written

1. [x] Trial -> Experimental Design -> Download Layout
2. [x] Trial -> Phenotype Summary Statistics -> Download Trial Data
3. [x] Wizard -> Related Trial Phenotypes
4. [x] ~Stock -> Traits Assayed -> Download Phenotypes~
    - This implementation is unnecessary. Missing measurements are not included for an additional stock download.
6. [x] Manage -> Trials -> Download Phenotypes

## Changes

<details>
  <summary><b>1. Trial -> Experimental Design -> Download Layout</b></summary>

This is my favorite menu for downloading trial data + phenotypes.
It has the greatest customizations for the layout, but no customization for the treatments and phenotypes.

image

</details>

<details>
  <summary><b>2. Trial -> Phenotype Summary Statistics -> Download Trial Data</b></summary>

I don't like this export format. It's extremely verbose, and displays trait columns in  a strange format (`SYNONYM = FULL TRAIT NAME|ID:00001`).

But I do really like the menu for selecting traits and trait components.
image
</details>

<details>
    <summary><b>3. Wizard -> Related Trial Phenotypes</b></summary>

This has the same export format as 2.

image
</details>

<details>
    <summary><b><s>4. Stock -> Traits Assayed -> Download Phenotypes</s></b></summary>
This UI interface actually doesn't need a missing data exported. Because any phenotypes that have not been recorded for an individual stock will not be exported at all.
Same export format as 2.

 image
</details>

<details>
    <summary><b>5. Manage -> Trials -> Download Phenotypes</b></summary>

Same export format as 2.
`mason/breeders_toolbox/trial/download_trials_phenotypes_dialog.mas` 

image
</details>

Checklist 
---------------------------------------------------------------------------------------
- [ ] Refactoring only
- [ ] Documentation only
- [ ] Fixture update only
- [x] Bug fix
  - [ ] The relevant issue has been closed.
  - [ ] Further work is required.
- [x] New feature
  - [x] Relevant tests have been created and run.
  - [ ] Data was added to the fixture
    - [ ] Data was added via a patch in `/t/data/fixture/patches/`.
  - [ ] User-Facing Change
    - [ ] The user manual in `/docs` has been updated.
  - [ ] Any new Perl has been documented using **perldoc**.
  - [ ] Any new JavaScript has been documented using **JSDoc**.
  - [ ] Any new _legacy_ JavaScript has been moved from `/js` to `/js/source/legacy`.

### CODE REVIEW

**Comment by ktmeaton** [comment]
I suspect I'm going to need to write quite a few new tests to cover this functionality.

**Comment by ktmeaton** [comment]
The BrAPI tests for the table of observationunits is failing. I think it might be because it needs missing values to be returned as undef. I can add that as an extra option in the backend.

**Comment by ktmeaton** [comment]
Failing test `t/selenium2/stock/stock_detail.t`:

```text
breedbase-1  | ok 18 - find delete synonym link
breedbase-1  | ok 19 - accept alert
breedbase-1  | not ok 20 - accept alert
breedbase-1  | ok 21 - get stock/38879/view test
breedbase-1  | ok 22 - find pedigree section
```

But this test passes locally, so I'm going to resubmit CI in case it's flaky.

File: lib/CXGN/BrAPI/v1/Studies.pm [hunk 1]
```diff
@@ -761,7 +761,8 @@ sub studies_table {
 		trial_list=>\@trial_ids_array,
 		trait_list=>\@trait_ids_array,
 		include_timestamp=>1,
-        exclude_phenotype_outlier=>$exclude_phenotype_outlier
+		exclude_phenotype_outlier=>$exclude_phenotype_outlier,
+		missing_format=>'undef',
```
Comment (ktmeaton) [inline lib/CXGN/BrAPI/v1/Studies.pm:765]: The BrAPI end points are unique in that they actually want missing data to be formatted as `undef`.

File: lib/CXGN/Phenotypes/Search/Native.pm [hunk 2]
```diff
@@ -348,16 +348,16 @@ sub search {
...
-                WHERE phenotype_id NOT IN (
+                WHERE phenotype_id IN (
                     SELECT phenotype_id
                     FROM phenotypeprop
                     WHERE type_id = $phenotype_outlier_type_id
                 )
-            ) AS not_outliers
-            ON not_outliers.phenotype_id = nd_experiment_phenotype.phenotype_id"
+            ) AS outliers
+            ON outliers.phenotype_id = nd_experiment_phenotype.phenotype_id"
```
Comment (ktmeaton) [inline lib/CXGN/Phenotypes/Search/Native.pm:360]: This is the fix for Issue #178 . The original logic had all records with missing data treated as phenotype outliers. I don't think this is desirable behavior, as a sample that doesn't have a measurement yet shouldn't be treated as a problematic one to exclude.

File: lib/CXGN/Phenotypes/Search/Native.pm [hunk 3]
```diff
@@ -614,6 +614,10 @@ sub search {
         OR observationunit.type_id = $tissue_sample_type_id)"; #plots AND plants AND subplots AND tissue_samples
     }
 
+    if ($self->exclude_phenotype_outlier) {
+        push @where_clause, "outliers.phenotype_id is NULL";
+    }
+
```
Comment (ktmeaton) [inline lib/CXGN/Phenotypes/Search/Native.pm:620]: More logic for #178 fix.

File: static/css/wizard.css [hunk 4]
```diff
@@ -138,3 +138,7 @@ div.wizard-panel .panel-body {
 .btn.wizard-btn-tag {
   pointer-events: none;
 }
+
+#wiz-down-1-c td, #wiz-down-3-c td {
+  white-space: normal;
+}
```
Comment (ktmeaton) [inline static/css/wizard.css:144]: Formatting to prevent bad overflow of wizard download components on small screen sizes. Allows line breaks between words in select tables.

File: t/lib/Shared/Genotypes.pm [hunk 5]
```diff
@@ -7,17 +7,19 @@ Shared::Genotypes - Shared functions for testing genotypes
...
 use LWP::UserAgent;
 use JSON;
 
-our @ISA= qw( Exporter );
-
-# Possible exports
-our @EXPORT_OK = qw( create_tissue_sample_genotypes upload_tissue_sample_genotypes $upload_tissue_sample_genotypes $vcf_expected_accessions $vcf_expected_tissue_samples );
-
 # Default exports
-our @EXPORT = qw( create_tissue_sample_genotypes upload_tissue_sample_genotypes $upload_tissue_sample_genotypes $vcf_expected_accessions $vcf_expected_tissue_samples );
```
Comment (ktmeaton) [inline t/lib/Shared/Genotypes.pm:20]: In writing the Phenotypes shared lib, I realized that configuring exports can be greatly simplified. I've updated the Genotypes lib to match.

File: t/lib/Shared/Genotypes.pm [hunk 6]
```diff
@@ -154,4 +156,26 @@ sub upload_tissue_sample_genotypes {
...
+        my $num_files = scalar(@downloaded_file_list);
+        $file_path = $num_files > 0 ? $downloaded_file_list[0] : undef;
+        $num_attempts += 1;
+        sleep(1);
+    }
+    if (defined($file_path)){
+        chomp($file_path);
+    }
+    return $file_path;
+}
```
Comment (ktmeaton) [inline t/lib/Shared/Genotypes.pm:181]: This function was originally defined in the Wizard search tests (`breeder_search.t`). Since this is a general utility function for genotype downloads, I've moved it into the shared library.

File: t/selenium2/breeders/breeder_search.t [hunk 7]
```diff
@@ -10,10 +10,13 @@ use Selenium::Firefox::Profile;
...
 use SGN::Test::Fixture;
+use Selenium::Firefox::Profile;
+use Shared::Phenotypes;
+use CXGN::Phenotypes::Missing;
+
 use JSON;
 use File::Slurp qw(read_file);
-use Data::Dumper;
-use Shared::Genotypes qw($vcf_expected_accessions $vcf_expected_tissue_samples create_tissue_sample_genotypes upload_tissue_sample_genotypes);
+use Shared::Genotypes;
```
Comment (ktmeaton) [inline t/selenium2/breeders/breeder_search.t:19]: Just import all the default functions and variables for brevity, rather than manually specifying everything individually. We also don't need `Data::Dumper` since that only provides debugging functionality.

File: t/selenium2/breeders/breeder_search.t [hunk 8]
```diff
@@ -497,15 +500,15 @@ $t->while_logged_in_as("submitter", sub {
     $t->click_ok('//span[text()="Related Genotype Data"]', 'xpath', 'open related genotype data panel');
     # save the current time, we will need this to find the correct downloaded file
     my $download_time = time();
-    $t->click_ok('wizard-download-genotypes', 'class', 'download genotypes');
+    $t->click_ok('//button[contains(@class, "wizard-download-genotypes")]', 'xpath', 'click download genotypes');
```
Comment (ktmeaton) [inline t/selenium2/breeders/breeder_search.t:503]: I really don't know how this selector could have worked previously. The genotype download button has multiple classes, so this selector is failing for me now because there is no element that just has the class `wizard-download-genotypes`. I truly don't know how this test could have passed in my previous PR...

File: t/selenium2/breeders/breeder_search.t [hunk 9]
```diff
@@ -517,7 +520,7 @@ $t->while_logged_in_as("submitter", sub {
     @vcf_response_observed = grep !/fileDate/, @vcf_response_observed;
     is_deeply(\@vcf_response_observed, \@vcf_response_expected, 'tissue sample vcf matches the expected values');
 
-    unlink($vcf_file_path) or die "Failed to delete $vcf_file_path: $!";
+    ok(unlink($vcf_file_path), 'delete downloaded vcf file');
```
Comment (ktmeaton) [inline t/selenium2/breeders/breeder_search.t]: Don't die, just fail the tests. We still want to know if the remaining tests pass.

File: t/selenium2/breeders/breeder_search.t [hunk 10]
```diff
@@ -537,15 +540,15 @@ $t->while_logged_in_as("submitter", sub {
     $t->click_ok('//span[text()="Related Genotype Data"]', 'xpath', 'open related genotype data panel');
     # save the current time, we will need this to find the correct downloaded file
     $download_time = time();
-    $t->click_ok('wizard-download-genotypes', 'class', 'download genotypes');
+    $t->click_ok('//button[contains(@class, "wizard-download-genotypes")]', 'xpath', 'click download genotypes');
```
Comment (ktmeaton) [inline t/selenium2/breeders/breeder_search.t:543]: Similar to above, this download button needs a more general class lookup.

File: t/selenium2/breeders/breeder_search.t [hunk 11]
```diff
@@ -557,32 +560,58 @@ $t->while_logged_in_as("submitter", sub {
...
-
-$t->driver()->quit();
+$t->driver->quit();
 $f->clean_up_db();
 done_testing();
-
-# Locate a file that is newer than the given timestamp.
-# Useful when downloading genotype data through the UI,
-# to find the correct output since it's dynamically named.
-sub get_file_newer_than_timestamp {
```
Comment (ktmeaton) [inline t/selenium2/breeders/breeder_search.t:571]: Moved to the shared lib file.

File: t/selenium2/breeders/breeder_search.t [hunk 12]
```diff
@@ -497,15 +500,15 @@ $t->while_logged_in_as("submitter", sub {
...
+    $t->click_ok('//button[contains(@class, "wizard-download-genotypes")]', 'xpath', 'click download genotypes');
 
     # search for a vcf file that is newer than before we clicked download
     my $vcf_file_path = get_file_newer_than_timestamp(
         "/selenium/downloads/breedbase_genotypes_*.vcf",
         $download_time,
         10 ,# maximum find attempts
     );
-    ok(defined($vcf_file_path), 'locate downloaded vcf file');
+    ok(defined($vcf_file_path) && -e $vcf_file_path, 'locate downloaded vcf file');
```
Comment (ktmeaton) [inline t/selenium2/breeders/breeder_search.t:511]: Being a little bit more cautious with this check, make sure that the variable is defined and that the file actually exists.

File: mason/breeders_toolbox/breeder_search_page.mas [hunk 13]
```diff
@@ -185,7 +185,7 @@ $dataset_id => undef
     </div>
   </div>
 
-  <div class="wizard-downloads panel-group col-sm-12 col-md-6" id="wiz-down-group" role="tablist" aria-multiselectable="true">
+  <div class="wizard-downloads panel-group col-sm-12 col-md-12 col-lg-6" id="wiz-down-group" role="tablist" aria-multiselectable="true">
```
Comment (ktmeaton) [inline mason/breeders_toolbox/breeder_search_page.mas:188]: Define better styles for this element on large screens.

File: t/selenium2/02_trial/trial_download.t [hunk 14]
```diff
@@ -146,8 +151,145 @@ $t->while_logged_in_as("curator", sub {
...
+            $phenotypes_download_path,                   # path file will be downloaded to
+        );
+        # Keep only the last three lines which are the header and data
+        my @observed = splice(@$observed, -3);
+        is_deeply(\@observed, $expected, 'download trials data file has expected values');
+    }
 });
 
-$t->driver->close();
+$t->driver->quit();
```
Comment (ktmeaton) [inline t/selenium2/02_trial/trial_download.t:293]: `close()` just closes a single window. `quit()` fully shuts down the driver. Since this is actually the end the tests, I think `quit()` makes more sense here.

File: t/selenium2/breeders/breeder_search.t [hunk 15]
```diff
@@ -517,7 +520,7 @@ $t->while_logged_in_as("submitter", sub {
     @vcf_response_observed = grep !/fileDate/, @vcf_response_observed;
     is_deeply(\@vcf_response_observed, \@vcf_response_expected, 'tissue sample vcf matches the expected values');
 
-    unlink($vcf_file_path) or die "Failed to delete $vcf_file_path: $!";
+    ok(unlink($vcf_file_path), 'delete downloaded accession vcf file');
```
Comment (ktmeaton) [inline t/selenium2/breeders/breeder_search.t:523]: Don't die, just fail the current test!

File: t/selenium2/breeders/breeder_search.t [hunk 16]
```diff
@@ -557,32 +560,58 @@ $t->while_logged_in_as("submitter", sub {
     @vcf_response_observed = grep !/fileDate/, @vcf_response_observed;
     is_deeply(\@vcf_response_observed, \@vcf_response_expected, 'tissue sample vcf matches the expected values');
 
+    ok(unlink($vcf_file_path), 'delete downloaded vcf file');
```
Comment (ktmeaton) [inline t/selenium2/breeders/breeder_search.t:563]: This test suite was previously missing it's cleanup file test.

**Comment by ktmeaton** [comment]
Something is happening between test 182 (Download missing format empty) and test 183 (download wizard has expected missing data empty). And it is severe enough that it actually forces the whole test suite to stop, which suggests it's an actual perl error. Maybe I'll enable stderr just to see what's happening.


```text
breedbase-1  | /home/production/cxgn/sgn/t/selenium2/breeders/breeder_search.r .................................. 
....
breedbase-1  | ok 180 - select download plot level
breedbase-1  | ok 181 - Select missing format empty
breedbase-1  | ok 182 - Download missing format empty
breedbase-1  | Dubious, test returned 2 (wstat 512, 0x200)
breedbase-1  | All 182 subtests passed 
breedbase-1  | /home/production/cxgn/sgn/t/selenium2/breeders/breeding_programs.t .................................. 
...
```

Edit1: After enabling stderr, the file is not downloading.

```text
breedbase-1  | TrialLayout generate layout Tue Jul  7 18:23:22 2026
breedbase-1  | ok 182 - Download missing format empty
breedbase-1  | EXP TYPE =field_layout
breedbase-1  | read_file '/selenium/downloads/Kasese_solgs_trial_phenotypes.csv' - open: No such file or directory at t/lib/Shared/Phenotypes.pm line 37.
breedbase-1  | NOTICE:  relation "logged_in_user" already exists, skipping
breedbase-1  | NOTICE:  relation "logged_in_user" already exists, skipping
breedbase-1  | wait_for_network_idle -> Active requests: 0
breedbase-1  | wait_for_network_idle -> Active requests: 0
breedbase-1  | wait_for_network_idle -> Network is now idle after 1 seconds of no active requests
breedbase-1  | ####DEMOLISHING THIS OBJECT NOW... AND GETTING DBSTATS AGAIN
breedbase-1  | Gathering DB stats...
```

- But we know this general functionality works in CI, because the tests in `trial_download.t` use the exact same function calls and download paths.
- Why would just this one test work locally but not in CI?

Edit2: The solution ended up being that it just needed more time to download. I upped the number of attempts from 10 (10 seconds) to 30 (30 seconds) and it now passes.

File: mason/breeders_toolbox/trial/download_layout_interface_dialog.mas [hunk 17]
```diff
@@ -106,6 +106,14 @@ $trial_stock_type => undef
                                 <input id="create_fieldbook_include_measured_<% $download_type %>" type="checkbox" data-toggle="toggle" />
                             </div>
                             <div class="layout-item">
+                                <label>Missing measurement format: </label>
+                                <select class="form-control" id="create_fieldbook_missing_format_<% $download_type %>">
+                                    <option value="empty">Empty Cell</option>
+                                    <option value="NA">NA</option>
+                                    <option value="period">.</option>
```
Comment (grapereader) [inline mason/breeders_toolbox/trial/download_layout_interface_dialog.mas]: Is 'undef' supposed to be an option here?

File: mason/breeders_toolbox/trial/download_trials_phenotypes_dialog.mas [hunk 18]
```diff
@@ -42,7 +42,18 @@
...
+            <div class="form-group">
+                <label class="col-sm-7 control-label">Missing measurement format: </label>
+                <div class="col-sm-5">
+                    <select class="form-control" id="download_trials_phenotypes_missing_format">
+                        <option value="empty">Empty Cell</option>
+                        <option value="NA">NA</option>
+                        <option value="period">.</option>
+                    </select>
+                </div>
+            </div>
```
Comment (grapereader) [inline mason/breeders_toolbox/trial/download_trials_phenotypes_dialog.mas:54]: Should this dropdown be centralized in its own component? What do you think?

File: mason/breeders_toolbox/trial/download_layout_interface_dialog.mas (hunk 17 above)
Comment (ktmeaton) [reply mason/breeders_toolbox/trial/download_layout_interface_dialog.mas]: The 'undef' option is really only meant for internal use in the BrAPI backend functions. These functions require missing data to be converted to undefined values. So the 'undef' option is not exposed to users when they're trying to export to a file through the UI.

File: mason/breeders_toolbox/trial/download_trials_phenotypes_dialog.mas (hunk 18 above)
Comment (ktmeaton) [reply mason/breeders_toolbox/trial/download_trials_phenotypes_dialog.mas:54]: Sure, I will put it into a reusable component!

File: t/selenium2/breeders/breeder_search.t (hunk 8 above)
Comment (grapereader) [reply t/selenium2/breeders/breeder_search.t:503]: Likely some flakiness going on, maybe related to opening the panel immediately before this.

The xpath is a bit more specific, only looking for buttons with the class, but the 2 methods should be otherwise identical. That button is the only element on the page with the "wizard-download-genotypes" class.

File: mason/breeders_toolbox/trial/download_trials_phenotypes_dialog.mas (hunk 18 above)
Comment (ktmeaton) [reply mason/breeders_toolbox/trial/download_trials_phenotypes_dialog.mas:54]: I've at least put the options in a reusable component, that pulls values directly from the perl module so they are only defined once in the code. This dropdown is used in 5 different pages, each with slightly different styling, classes and ids for the select element. So I've just made the options reusable.

Comment (grapereader) [reply mason/breeders_toolbox/trial/download_trials_phenotypes_dialog.mas:54]: Nice that is a good compromise. This looks clean! 🥳

**Review by grapereader — APPROVED** [review]
Looks great to me! 👍
